main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification.scala

```diff
@@ -127,26 +129,100 @@ object EffUnification {
 
     }
 
-    // Give up early if either type contains an associated type.
-    if (Type.hasAssocType(tpe1) || Type.hasAssocType(tpe2)) {
-      return Ok((Substitution.empty, List(Ast.BroadEqualityConstraint(tpe1, tpe2))))
+    //
+    // We handle associated types by:
+    // 1. replacing each associated type with a fresh rigid variable
+    // 2. performing unification
+    // 3. replacing the variables with the types they represent
+    //    - we do this by composing a substitution back to the associated types
+    //      with the substitution from unification
+    //
+    // First clear any associated types from the types, temporarily replacing simple associated types with variables
+    clearAssocs(tpe1, tpe2, renv0) match {
+      // Case 1: There were non-reduced associated types. Give up and defer them.
+      case None => Ok((Substitution.empty, List(Ast.BroadEqualityConstraint(tpe1, tpe2))))
+
+      // Case 2: All associated types were reduced. Continue with unification.
+      case Some((t1, t2, back)) =>
+
+        // add the associated type marker variables to the rigidity environment
+        val renv = back.keys.foldLeft(renv0)(_.markRigid(_))
+
+        // build a substitution from the variable mapping
+        val backSubst = back.map {
+          case (sym, (assoc, tvar)) =>
+            val tpe = Type.AssocType(Ast.AssocTypeConstructor(assoc, SourceLocation.Unknown), Type.Var(tvar, SourceLocation.Unknown), sym.kind, SourceLocation.Unknown)
+            sym -> tpe
+        }
+
+        // Choose the SVE implementation based on the number of variables.
+        val numberOfVars = (t1.typeVars ++ t2.typeVars).size
+        val threshold = flix.options.xbddthreshold.getOrElse(DefaultThreshold)
+
+        val substRes = if (numberOfVars < threshold) {
+          implicit val alg: BoolAlg[BoolFormula] = new BoolFormulaAlg
+          implicit val cache: UnificationCache[BoolFormula] = UnificationCache.GlobalBool
+          lookupOrSolve(t1, t2, renv)
+        } else {
+          implicit val alg: BoolAlg[DD] = new BddFormulaAlg
+          implicit val cache: UnificationCache[DD] = UnificationCache.GlobalBdd
+          lookupOrSolve(t1, t2, renv)
+        }
+
+        substRes.map {
+          case subst0 =>
+            // replace the temporary variables with the assoc types they represent
+            val subst1 = Substitution(backSubst) @@ subst0
+            // remove the temporary variables from the substitution
+            val subst = backSubst.keys.foldLeft(subst1)(_.unbind(_))
+            (subst, Nil)
```

Comment:
Would we want to remove some bindings from subst? (e.g. the bindings for variables we no longer care about).

Reply:
Isn't that what line 177 does?

Reply:
Aaah. Something about forests and trees.